### PR TITLE
Use one SQL transaction per Datastore transaction in replay to SQL

### PR DESCRIPTION
**Pull request by gbrodman**
There was a subtle issue that we encountered in sandbox when using one
transaction per file that was difficult to replicate. Basically,
1. Save a domain with dsData
2. Save the domain without dsData
3. Save the domain with the same dsData as step 1
4. Delete literally any object

If one performs steps 2-4 in the same transaction, Hibernate will throw
an exception (cascade re-saving a cascade-deleted object). Note that
step 4 is in fact necessary to reproduce the issue, yay Hibernate.

We will test this and if one transaction per transaction is too slow,
we'll figure out ways to reduce the number of SQL transactions.

---
This change is [Reviewable](https://reviewable.io/reviews/google/nomulus/1268)

### CODE REVIEW

**Review by CydeWeys — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1268)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @gbrodman)

___
*[core/src/main/java/google/registry/backup/ReplayCommitLogsToSqlAction.java, line 211 at r1](https://reviewable.io/reviews/google/nomulus/1268#-MgMGCRNAQHqE-iwqWg1:-MgMGCRNAQHqE-iwqWg2:bhnqcxo) ([raw file](https://github.com/google/nomulus/blob/2608a276c12ca2f4a708d2a9500ffec26f302d99/core/src/main/java/google/registry/backup/ReplayCommitLogsToSqlAction.java#L211)):*
> ```Java
>       // if we succeeded, set the last-seen time
>       DateTime checkpoint = DateTime.parse(metadata.getName().substring(DIFF_FILE_PREFIX.length()));
>       jpaTm().transact(() -> SqlReplayCheckpoint.set(checkpoint));
> ```

This doesn't work.  This needs to be re-saved inside the transaction below.  Yes, that's going to involve a lot of churn, but there's no other way to do it and be safe.

**Review by gbrodman — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1268)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @CydeWeys)

___
*[core/src/main/java/google/registry/backup/ReplayCommitLogsToSqlAction.java, line 211 at r1](https://reviewable.io/reviews/google/nomulus/1268#-MgMGCRNAQHqE-iwqWg1:-MgMaOOnBw1M1Xftdc7m:b9pql5t) ([raw file](https://github.com/google/nomulus/blob/2608a276c12ca2f4a708d2a9500ffec26f302d99/core/src/main/java/google/registry/backup/ReplayCommitLogsToSqlAction.java#L211)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

This doesn't work.  This needs to be re-saved inside the transaction below.  Yes, that's going to involve a lot of churn, but there's no other way to do it and be safe.
</blockquote></details>

Fair point, done.

**Review by CydeWeys — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1268)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @gbrodman)

___
*[core/src/main/java/google/registry/backup/ReplayCommitLogsToSqlAction.java, line 215 at r2](https://reviewable.io/reviews/google/nomulus/1268#-MgR9TdzB_qnNusaK9AB:-MgR9TdzB_qnNusaK9AC:bk0x5cn) ([raw file](https://github.com/google/nomulus/blob/493b8f1d33c916c92a7bd1be8a84ccfbde55e263/core/src/main/java/google/registry/backup/ReplayCommitLogsToSqlAction.java#L215)):*
> ```Java
>                   // if we succeeded, set the last-seen time
>                   DateTime checkpoint =
>                       DateTime.parse(metadata.getName().substring(DIFF_FILE_PREFIX.length()));
> ```

Wait, the checkpoint isn't the specific transaction within the file, it's just the file overall?  How does that work?  Won't this be doing a lot of noops then?

What happens if a file is partially committed, then the task fails, and re-runs 3 minutes later?  How does it successfully resume from the middle of the file?  Won't the earlier transactions that have already been replayed fail?

**Review by CydeWeys — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1268)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @gbrodman)

___
*[core/src/main/java/google/registry/backup/ReplayCommitLogsToSqlAction.java, line 215 at r2](https://reviewable.io/reviews/google/nomulus/1268#-MgR9TdzB_qnNusaK9AB:-MgRABgdDAsG--iW7XvV:b-6ltqmo) ([raw file](https://github.com/google/nomulus/blob/493b8f1d33c916c92a7bd1be8a84ccfbde55e263/core/src/main/java/google/registry/backup/ReplayCommitLogsToSqlAction.java#L215)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

Wait, the checkpoint isn't the specific transaction within the file, it's just the file overall?  How does that work?  Won't this be doing a lot of noops then?

What happens if a file is partially committed, then the task fails, and re-runs 3 minutes later?  How does it successfully resume from the middle of the file?  Won't the earlier transactions that have already been replayed fail?
</blockquote></details>

If the checkpoint is on a per-file basis, it wouldn't make sense to keep updating it after every transaction, since most of those are no-ops.  You'd only update it once the entire file has committed and you're moving onto the next one.  And then you also need a way to track progress within the transactions in the file, which doesn't appear to be present.  You're using the export file datetime as the checkpoint, but we also need the transaction times (or indexes) of the transactions within those files, no?

**Review by gbrodman — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1268)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @CydeWeys)

___
*[core/src/main/java/google/registry/backup/ReplayCommitLogsToSqlAction.java, line 215 at r2](https://reviewable.io/reviews/google/nomulus/1268#-MgR9TdzB_qnNusaK9AB:-MgRBOcPAzFYDN7yxUHs:bo9cvcc) ([raw file](https://github.com/google/nomulus/blob/493b8f1d33c916c92a7bd1be8a84ccfbde55e263/core/src/main/java/google/registry/backup/ReplayCommitLogsToSqlAction.java#L215)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

If the checkpoint is on a per-file basis, it wouldn't make sense to keep updating it after every transaction, since most of those are no-ops.  You'd only update it once the entire file has committed and you're moving onto the next one.  And then you also need a way to track progress within the transactions in the file, which doesn't appear to be present.  You're using the export file datetime as the checkpoint, but we also need the transaction times (or indexes) of the transactions within those files, no?
</blockquote></details>

That's why the checkpoint was originally per-file, and it probably still could be per-file. We don't anticipate that replaying the same transactions (from the first half of the file before it failed) would cause any issues because the only situation that would cause issues would involve hard-deleted entities that other entities had foreign keys on -- we discussed this if you recall, and basically only poll messages are hard-deleted.

**Review by CydeWeys — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1268)* status: 0 of 2 files reviewed, all discussions resolved (waiting on @gbrodman)

___
*[core/src/main/java/google/registry/backup/ReplayCommitLogsToSqlAction.java, line 215 at r2](https://reviewable.io/reviews/google/nomulus/1268#-MgR9TdzB_qnNusaK9AB:-MgRECss9uzxqz3sr1Ry:b-jsdu2m) ([raw file](https://github.com/google/nomulus/blob/493b8f1d33c916c92a7bd1be8a84ccfbde55e263/core/src/main/java/google/registry/backup/ReplayCommitLogsToSqlAction.java#L215)):*
<details><summary><i>Previously, gbrodman wrote…</i></summary><blockquote>

That's why the checkpoint was originally per-file, and it probably still could be per-file. We don't anticipate that replaying the same transactions (from the first half of the file before it failed) would cause any issues because the only situation that would cause issues would involve hard-deleted entities that other entities had foreign keys on -- we discussed this if you recall, and basically only poll messages are hard-deleted. 
</blockquote></details>

Ok, so if we're not checkpointing at the file level then, and replay of transactions that have already been successfully committed is OK, then the checkpoint should only be updated at the end of processing the entire file, as it had been previously.

But what about the situation where we have something like "update A with state 1351, update A with state 1240, update A with state 2629", etc., in a single file?  E.g. corresponding to a rapid series of domain update operations?  We're gonna be doing some unnecessary reversions and then re-reversions if we end up replaying most of that file, right?  But the end state is guaranteed to be the same?

Also, separately, I just want to confirm again for safety at what point the eventual consistency from Datastore is converted into strongly ordered transactional consistency with a guaranteed monotonically increasing tx time.  Is that guaranteed true of these commit log export files?  Every single TX within them has a strictly increasing Datastore transaction timestamp?  And there's no situation where there can be transactions in file A that have a later transaction time than any transaction in file A+3?  (Where A+3 is the chronologically next commit log export file.)

**Review by gbrodman — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1268)* status: 0 of 2 files reviewed, all discussions resolved (waiting on @gbrodman)

___
*[core/src/main/java/google/registry/backup/ReplayCommitLogsToSqlAction.java, line 215 at r2](https://reviewable.io/reviews/google/nomulus/1268#-MgR9TdzB_qnNusaK9AB:-MgRVGUA7yHeywSuM2VI:bgkymdu) ([raw file](https://github.com/google/nomulus/blob/493b8f1d33c916c92a7bd1be8a84ccfbde55e263/core/src/main/java/google/registry/backup/ReplayCommitLogsToSqlAction.java#L215)):*
> then the checkpoint should only be updated at the end of processing the entire file, as it had been previously.

Done

> We're gonna be doing some unnecessary reversions and then re-reversions if we end up replaying most of that file, right? But the end state is guaranteed to be the same?

Yes to both. Because the objects being replayed are objects and not only diffs, we should be ok. 

> eventual consistency

Take a look at the comment at https://github.com/google/nomulus/blob/master/core/src/main/java/google/registry/backup/CommitLogCheckpointStrategy.java#L45-L51

tldr is that any transaction that depends on another transaction will have a later timestamp (mentioned in that file, but not implemented in that file) and that commit logs are bunched together in files such that for each commit log in a file, the file will also contain commit logs that (were written before that commit log && have an earlier timestamp than that commit log) 

I think that should answer your question.

**Review by CydeWeys — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1268)* status: 0 of 2 files reviewed, all discussions resolved (waiting on @gbrodman)
